Homework7

matua opened the pull request:
https://go.skillbox.ru/education/course/java-framework-spring/8b55d8b6-baaa-4dc5-9385-9ecd4a83e4d4/homework

Commit:
init postgresql setup

Commit:
Java Persistance API

Commit:
Hibernate

Commit:
Spring Data JPA

Commit:
Переписать AuthorService с использованием JPA 🦾

Проведите рефакторинг AuthorService с целью перехода на более высокоуровневый интерфейс JPA. Мы уже переписали BookService с использованием возможностей JPA, но AuthorService всё ещё реализован на Spring JDBC.

Commit:
Переписать AuthorService с использованием JPA 🦾

Проведите рефакторинг AuthorService с целью перехода на более высокоуровневый интерфейс JPA. Мы уже переписали BookService с использованием возможностей JPA, но AuthorService всё ещё реализован на Spring JDBC.

Commit:
Merge remote-tracking branch 'origin/Homework5.7' into Homework5.7

# Conflicts:
#	src/main/java/com/example/mybookshopapp2/data/AuthorRepository.java

Commit:
Спроектировать структуру данных проекта 🦾🦾

Проведите анализ информации, отображаемой фронтендом, и спроектируйте структуру данных вашего интернет-магазина при помощи JPA. Не переживайте, что можете учесть не все детали.

Текущего набора сущностей, состоящего из Author и Book, уже недостаточно для отражения всей информации, необходимой фронтенду. У авторов есть биографии; у книг есть цена, названия, рейтинги, теги, рецензии и описания. Кроме книг и авторов, ещё есть покупатели и их учётные записи.

Commit:
Спроектировать структуру данных проекта

1. added Users table sql fill-up

Commit:
Спроектировать структуру данных проекта

worked on join tables

Commit:
Спроектировать структуру данных проекта

added repositories and services for all the entities

Commit:
Спроектировать структуру данных проекта

updated GenresController

Commit:
Спроектировать структуру данных проекта

added a couple more Controllelrs

Commit:
Спроектировать структуру данных проекта

added a couple more Controllers - fixes

Commit:
Подключить Liquibase к проекту интернет-магазина 🦾🦾🦾

Подключите к проекту интернет-магазина систему Liquibase для управления миграциями схем базы данных.

В примере мы подключили Liquibase к небольшому тестовому проекту. Сначала  сложно оценить выгоду от поддержки версионности схем базы данных проекта. Но в процессе разработки структура данных будет неизбежно меняться, и необходима возможность откатить изменения и начать заново с последнего рабочего варианта.

Commit:
Подключить Liquibase к проекту интернет-магазина 🦾🦾🦾

Подключите к проекту интернет-магазина систему Liquibase для управления миграциями схем базы данных.

В примере мы подключили Liquibase к небольшому тестовому проекту. Сначала  сложно оценить выгоду от поддержки версионности схем базы данных проекта. Но в процессе разработки структура данных будет неизбежно меняться, и необходима возможность откатить изменения и начать заново с последнего рабочего варианта.

Commit:
Подключить Liquibase к проекту интернет-магазина 🦾🦾🦾

Подключите к проекту интернет-магазина систему Liquibase для управления миграциями схем базы данных.

В примере мы подключили Liquibase к небольшому тестовому проекту. Сначала  сложно оценить выгоду от поддержки версионности схем базы данных проекта. Но в процессе разработки структура данных будет неизбежно меняться, и необходима возможность откатить изменения и начать заново с последнего рабочего варианта.

Commit:
Подключить Liquibase к проекту интернет-магазина 🦾🦾🦾

Подключите к проекту интернет-магазина систему Liquibase для управления миграциями схем базы данных.

В примере мы подключили Liquibase к небольшому тестовому проекту. Сначала  сложно оценить выгоду от поддержки версионности схем базы данных проекта. Но в процессе разработки структура данных будет неизбежно меняться, и необходима возможность откатить изменения и начать заново с последнего рабочего варианта.

Commit:
Refactoring

Commit:
init

Commit:
REST API Documentation

Commit:
REST API Creation

Commit:
Spring REST Pagination

Commit:
Подключить постраничный вывод для ленты «Новинки» и «Популярное» на главной странице 🦾

Теперь, когда лента рекомендуемых книг на главной странице снабжена механизмом постраничного вывода загружаемых из базы данных книг, реализуйте такое же поведение для лент «Новинки» и «Популярное». Сделайте всю необходимую логику и структуры для работы этих лент.

Commit:
Подключить постраничный вывод для ленты «Новинки» и «Популярное» на главной странице

Теперь, когда лента рекомендуемых книг на главной странице снабжена механизмом постраничного вывода загружаемых из базы данных книг, реализуйте такое же поведение для лент «Новинки» и «Популярное». Сделайте всю необходимую логику и структуры для работы этих лент.

Commit:
Реализовать список книг:
по новинкам 🦾🦾
по популярным 🦾🦾

Commit:
Реализовать список книг:
по новинкам 🦾🦾
по популярным 🦾🦾

BUG FIXES

Commit:
Реализовать список книг:
по новинкам
по популярным

BUG FIXES -2

Commit:
Реализовать список книг:
по новинкам
по популярным

BUG FIXES -3

Commit:
Реализовать список книг:
по новинкам
по популярным

BUG FIXES -4

Commit:
Реализовать список книг:
по новинкам
по популярным

BUG FIXES -5

Commit:
Реализовать список книг:
по тегам 🦾🦾🦾

entity and db tables set up

Commit:
lombok incorporation

Commit:
main page, recent page, popullar page pagination bug fixes on offset and limit

Commit:
genres fixed

Commit:
Реализовать список книг:
по тегам 🦾🦾🦾

Commit:
tags - fixes

Commit:
general - fixes

Commit:
on the way to author page

Commit:
lombok @ManyToMany bug fix

@EqualsAndHashCode(exclude = "books")
@ToString(exclude = "books")

Commit:
tags bug fix

Commit:
tags bug fix - 2

Commit:
tags bug fix - 3

Commit:
author page

Commit:
Реализовать страницу автора 🦾🦾🦾🦾

У каждого автора в интернет-магазине книг есть личная страница. Вы можете перейти на неё из раздела «Авторы», найдя писателя в алфавитном списке. На личной странице автора есть фотография и краткая биография. Внизу страницы показан список книг, принадлежащих автору. Список нужно вывести постранично, на случай, если книг будет много. Реализуйте страницу автора, всю необходимую логику и структуры для работы этого раздела магазина.

Commit:
author page bug fixes

Commit:
author page bug fixes - 2

Commit:
refactoring and bug fixes

Commit:
refactoring and bug fixes of fetching books and wrong display of authors

Commit:
refactoring and bug fixes of fetching books and wrong display of authors - 2

Commit:
refactoring and bug fixes of fetching books and wrong display of authors - 3

Commit:
refactoring and bug fixes of fetching books and wrong display of authors - 3

Commit:
video 7.1

Commit:
video 7.1 - 2

Commit:
video 7.2 - Books downloading

Commit:
video 7.3 - Exceptions handling

Commit:
video 7.4 - bug fixes

Commit:
video 7.4

matua committed:
Bug fixes in JS links to books (slug)

matua committed:
Implemented custom serializer for Book entity

matua committed:
Implemented custom serializer for Book entity - 2

matua committed:
Implemented custom serializer for Book entity - 2

matua committed:
1. Implemented custom serializer for Book entity - 3
2. Bug fix on JS author(s) output

matua committed:
polishing cart.h

matua committed:
1. GlobalControllerAdvice
2. partial postponed logic implementation

matua committed:
Реализовать добавление книг в категорию «Отложенное» 🦾

matua committed:
Закрепить знания, полученные в седьмом модуле. степень сложности задания.&nbsp

matua committed:
Закрепить знания, полученные в седьмом модуле. степень сложности задания.&nbsp - 2
(adding global access to cookies of cart and postponed via @ModelAttribute and @GlobalAdvice)

matua committed:
Закрепить знания, полученные в седьмом модуле. степень сложности задания.&nbsp - 2
(adding global access to cookies of cart and postponed via @ModelAttribute and @GlobalAdvice)

Commit:
Spring Hateoas added, Book extended with RepresentationModel

Commit:
Merge remote-tracking branch 'origin/Homework7' into Homework7

matua committed:
1. Внедрить HATEOAS в рамках BooksRestApiController
2. Removed custom serialization, and switched to @JsonIdentityInfo

matua committed:
Merge remote-tracking branch 'origin/Homework7' into Homework7

# Conflicts:
#	pom.xml
#	src/main/java/com/example/mybookshopapp2/model/Book.java

matua committed:
1. Внедрить HATEOAS в рамках BooksRestApiController - changed Link creation to Method instance handed to WebMvcLinkBuilder

matua committed:
1. Внедрить HATEOAS в рамках BooksRestApiController - 2

matua committed:
1. Внедрить HATEOAS в рамках BooksRestApiController - 3

matua committed:
1. Внедрить HATEOAS в рамках BooksRestApiController - 3

matua committed:
refactor

matua committed:
correcting error on bookid passed for rating

matua committed:
book_rating jpa saving implementation

matua committed:
js update

matua committed:
js update - push bug search

matua committed:
JS script fixed

matua committed:
getBookRatingMap implementation

matua committed:
getTotalNumberOfRatings implementation

matua committed:
getAverageRating implementation

matua committed:
BookReview implementation - part 1

matua committed:
BookReview implementation - part 2 (bug with @ModelAttribute fixed)

matua committed:
BookReview implementation - part 3

matua committed:
BookReview implementation - part 4

matua committed:
Adding from postponed to cart

matua committed:
Fixed the reload page upon thumbing up/down of the book reviews

caligula95 commented on pom.xml:
```diff
@@ -25,21 +25,61 @@
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-web</artifactId>
         </dependency>
+<!--        <dependency>-->
```
remove the code if you don't use it. Comments are for commenting but not for the code hiding.

caligula95 commented on src/main/java/com/example/mybookshopapp2/config/SpringfoxConfig.java:
```diff
@@ -0,0 +1,39 @@
+        return new ApiInfo(
+                "Bookshop API", "API for bookstore", "1.0",
+                "https://termsofservice.org",
+                new Contact("API owner", "https://www.ownerssite.com", "owner@rmailer.com"),
```
Please put correct values for the contact. owner@rmailer.com is not your email. If you don't want to put the values - remove the property, but don't add placeholders

caligula95 commented on src/main/java/com/example/mybookshopapp2/controllers/AuthorController.java:
```diff
@@ -26,9 +32,30 @@ public AuthorController(AuthorService authorService) {
-    public String authorPage() {
-        return "/authors/slug";
+    @ModelAttribute("author")
+    public Author getAuthor(@PathVariable("author") String author) {
```
please add some logs to log requests and responses. It will help you to debug the application.

caligula95 commented on src/main/java/com/example/mybookshopapp2/controllers/BooksController.java:
```diff
@@ -0,0 +1,167 @@
+    private final BookReviewLikeService bookReviewLikeService;
+
+    @Autowired
+    public BooksController(BookRepository bookRepository, BookRatingService bookRatingService, ResourceStorage storage, BookRatingRepository bookRatingRepository, BookReviewService bookReviewService, BookReviewLikeService bookReviewLikeService) {
```
Why are you using repository in controller? You should use service instead of repository. Layer decoupling will help you to organise your code with loose coupling.

caligula95 commented on src/main/java/com/example/mybookshopapp2/controllers/BooksController.java:
```diff
@@ -0,0 +1,167 @@
+                             @RequestParam String bookSlug,
+                             @RequestParam String text) {
+        BookReview bookReview = new BookReview();
+        User user = new User(); //temp for testing
```
It's better to specify some separate class for creating temp data. Don't mix this stuff in the controllers. Remember about single responsibility.

caligula95 commented on src/main/java/com/example/mybookshopapp2/controllers/BooksController.java:
```diff
@@ -0,0 +1,167 @@
+    public ResponseEntity<ByteArrayResource> bookFile(@PathVariable("hash") String hash) throws IOException {
+
+        Path path = storage.getBookFilePath(hash);
+        Logger.getLogger(this.getClass().getSimpleName()).info("book file path: " + path);
```
You can create a global log variable to log all the requests and responses in controller.

caligula95 commented on src/main/java/com/example/mybookshopapp2/controllers/BooksController.java:
```diff
@@ -0,0 +1,167 @@
+
+    private List<BookReview> getBookReviewsByBookId(@PathVariable("slug") String slug) {
+        Integer bookId = bookRepository.findBookBySlug(slug).getId();
+        return bookReviewService.getBookReviewsByBookId(bookId);
```
This code should be in service or component.

caligula95 commented on src/main/java/com/example/mybookshopapp2/controllers/BooksRestApiController.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+
+    @ExceptionHandler(MissingServletRequestParameterException.class)
+    public ResponseEntity<ApiResponse<Book>> handleMissingServletRequestParameterException(Exception exception) {
+        return new ResponseEntity<>(new ApiResponse<>(HttpStatus.BAD_REQUEST, "Missing required parameters",
+                exception), HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(BookstoreApiWrongParameterException.class)
+    public ResponseEntity<ApiResponse<Book>> handleBookstoreApiWrongParameterException(Exception exception) {
+        return new ResponseEntity<>(new ApiResponse<>(HttpStatus.BAD_REQUEST, "Bad parameter value...", exception)
+                , HttpStatus.BAD_REQUEST);
+    }
```
Please put exception handlers into the separate class

caligula95 commented on src/main/java/com/example/mybookshopapp2/controllers/BooksRestApiController.java:
```diff
@@ -0,0 +1,122 @@
+        return ResponseEntity.ok(bookService.getBooksWithPriceBetween(min, max));
+    }
+
+    @GetMapping("/with-max-discount")
```
Please stick to the single code style and naming. Previous methods you called with by prefix. by-price.. , by-title etc. But this you called with-max-discount. Isn't it's better to call it by-max-discount?

matua replied:
This code was written by the professor of the video course.

caligula95 commented on src/main/java/com/example/mybookshopapp2/controllers/CartController.java:
```diff
@@ -1,34 +1,87 @@
-    public CartController(AuthorService authorService) {
-        this.authorService = authorService;
+    public CartController(BookRepository bookRepository) {
+        this.bookRepository = bookRepository;
```
Don't use repo in controllers.

caligula95 commented on src/main/java/com/example/mybookshopapp2/controllers/GenresController.java:
```diff
@@ -1,29 +1,36 @@
 import java.util.Map;
 
 @Controller
 public class GenresController {
```
I'm not sure this is valid controller. RequestMapping is missing

matua replied:
```
@GetMapping("/genres")
    public String genresPage() {
        logger.debug("Rendering genres/index.html");
        return "/genres/index";
    }
```

caligula95 commented on src/main/java/com/example/mybookshopapp2/model/Author.java:
```diff
@@ -0,0 +1,48 @@
+@Entity
+@Table(name = "authors")
+@ApiModel(description = "data model of author entity")
+@JsonIdentityInfo(
```
It's not good to use the same class object for the db model and the response in controllers. It's better to use DTO classes for the controllers. You should not mix persistence annotations with json and swagger annotations.

matua committed:
Please put correct values for the contact. owner@rmailer.com is not your email. If you don't want to put the values - remove the property, but don't add placeholders

matua committed:
please add some logs to log requests and responses. It will help you to debug the application.

matua committed:
Why are you using repository in controller? You should use service instead of repository. Layer decoupling will help you to organise your code with loose coupling.

matua committed:
It's better to specify some separate class for creating temp data. Don't mix this stuff in the controllers. Remember about single responsibility.

Why are you using repository in controller? You should use service instead of repository. Layer decoupling will help you to organise your code with loose coupling.

matua committed:
This code should be in service or component.

matua committed:
Don't use repo in controllers.

matua committed:
Code comments cleanup
